x-pack/plugins/upgrade_assistant/server/routes/node_disk_space.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ClusterGetSettingsResponse } from '@elastic/elasticsearch/api/types';
+import { ByteSizeValue } from '@kbn/config-schema';
+import { API_BASE_PATH } from '../../common/constants';
+import { versionCheckHandlerWrapper } from '../lib/es_version_precheck';
+import { RouteDependencies } from '../types';
+
+interface NodeWithLowDiskSpace {
+  nodeId: string;
+  nodeName: string;
+  used: string;
+  lowDiskWatermarkSetting: string;
+}
+
+// Buffer percentage used to determine if the user has low disk space
+// For example, if a user has used 80%, and has defined a low disk watermark setting of 85%, we would warn the user
+const LOW_DISK_SPACE_BUFFER_PERCENTAGE = 5;
```

Comment:
@masseyke can you provide feedback on what a good number might be to use here to show a warning? Currently, if the disk space is 5% less than the low disk watermark setting, UA displays a warning.

Reply:
Do we need a buffer? It's OK to briefly exceed the low watermark during a reindex right?

Reply:
@sebelga can you share your thoughts on what you were originally thinking here?

If we only want to display a warning if the low watermark setting is hit, then perhaps we don't need to do this logic on our side at all, and can determine whether or not to show the warning based on whether the deprecation implemented in https://github.com/elastic/elasticsearch/issues/82807 is present.

Reply:
> It's OK to briefly exceed the low watermark during a reindex right?

I don't know if it's OK or not 😊  that's why I thought adding a buffer was safer.

We anyway need _some_ buffer. If the user is already above or at the low watermark ES will return a critical deprecation (added in https://github.com/elastic/elasticsearch/pull/83544) and won't allow to upgrade. In that scenario, instead of a callout warning, shouldn't we prevent the reindexing altogether?

Reply:
The upgrade assistant will report a critical message if a node is over the high water mark. I thought that this was an additional protection in case something filled up a disk between when the upgrade assistant called the deprecation info API and when the user clicked the button to reindex an index. If that's the case, we don't need additional buffer -- just the same check that the deprecation info API is doing. It's definitely possible that I'm misunderstanding the purpose of this check though.

Reply:
> Do we need a buffer? It's OK to briefly exceed the low watermark during a reindex right?

> I don't know if it's OK or not 😊 that's why I thought adding a buffer was safer.

Based on these comments, I would like to get a better understanding of what is safe to assume 😄 . @masseyke is there someone on the ES team who might be able to provide some additional guidance?

Reply:
Here are our docs for the 3 levels: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cluster.html#disk-based-shard-allocation. Once the low watermark is exceeded, we won't allocate replica shards but we will allocate the primaries. So if all nodes exceeded the low watermark and somehow didn't come back down after the reindex then the cluster would be in a yellow state (but it would succeed).

Reply:
> Based on these comments, I would like to get a better understanding of what is safe to assume 😄 . @masseyke is there someone on the ES team who might be able to provide some additional guidance?

I'm sure @dakrone could chime in here. It sounds like this is just a question of what we want the kibana user experience to be though right? It would be kind of odd to have them set a low, high, and flood_stage watermark, and then tell them that they can't reindex because of a 4th implicit watermark, right?

Reply:
OK, thanks @masseyke! It sounds like the logic I've added here isn't necessary then, and we can use the check the deprecation info API is doing. 

@sebelga does that sound OK with you, or is there something we're still not considering? 

>  In that scenario, instead of a callout warning, shouldn't we prevent the reindexing altogether?

Based on Keith's latest information, I'm not sure if we should block reindexing.

Reply:
@alisonelizabeth I think we should also gain an understanding of how autoscaling affects things. If autoscaling renders concerns about running out space moot, then that's another argument to not block reindexing but to rather show the user a warning.

Reply:
@masseyke are you able to confirm whether or not Seb's scenario could occur? Or is there someone else I should reach out to regarding this? Happy to jump on a quick call if that's easier.

Reply:
> Does that mean calling the deprecation API before each reindex op and parse the deprecations to see if one of them is related to low disk space?

@sebelga that was my initial thought. I _think_ we could leverage the initial call to the deprecation info API (used to render the table) and pass that info to the flyout when it is opened vs. calling the API every time the reindex flyout is opened. Not ideal, as then we might have to rely on the message string or ask the ES team to add some data to the `_meta` field of the deprecation. 

We could also keep the logic I already have here (without the buffer). I would just want to make sure there isn't the possibility that the UI would ever get out of sync with ES.

Reply:
There's always the chance that a customer will have a tiny cluster with small disks and a very large index to reindex. I think that's the case we're worried about right? In a normal cluster with normal indexes, doing a reindex of a single index is not likely to shoot past both the low watermark and the high watermark. And if a single reindex can do that on a small cluster, I think they probably ought to have set a lower low watermark in the first place because of that risk.
Maybe what you really want to do is to check the size of the index that you're about to reindex, and see the nodes have enough capacity to temporarily absorb that size again without hitting the high water mark? Even then, something could come along and fill up a disk (for example a node goes down and all its shards are pushed out to other nodes; or some non-Elastic process we know nothing about fills up a disk). Without some big changes to Elasticsearch, I don't think there's a way to 100% guarantee that reindex will complete once it has started.
I'd be glad to get on a call if that would help. And I'm definitely not a reindex expert, so I'd probably ask @dakrone for advice there. I did look in the code though, and reindex is not doing anything to check watermarks.

Reply:
+1 to what Keith said, though I would add that the chance of it affecting the reindex is small, for example, these steps:

> 1. The user is below all watermarks. So green to go and UA does not warn of anything
> 2. They start a reindex op
> 3. ES throws an error in the middle because of lack of disk space (the error is quite technical and not clear why something failed)
> 4. The user is looking at the big red error and wonders "What just happened here? I can't reindex, I need to contact support"

ES will *only* throw an error in the middle if the flood stage has been reached. Otherwise during the reindex after the shards have already been allocated (which they will be when the new index is first created) there won't be any errors. If they are already over the low watermark, then the initial allocation of the replicas may fail. If they are already over the high watermark, then the initial creation of the index will fail and the reindex will fail. I don't know how many concurrent reindexes the UI will kick off (I hope it doesn't try to concurrently reindex every single index at once). However, the chance of the initial allocations failing is smaller considering the index will be 0b when first created.

With that being said, I don't think we should prevent reindex (which I think was agreed on in this thread already?), but a reasonable warning seems appropriate. I agree with Keith that there is no way to guarantee a reindex will complete once it's started, however autoscaling will alleviate this to a great degree on Cloud. So, I think the existing warning at 5% is fine.

Reply:
Thanks for your insight @dakrone 👍 

> I hope it doesn't try to concurrently reindex every single index at once

@alisonelizabeth does UA prevent in anyway the user to do that?

Reply:
@sebelga There isn't the ability to select multiple indices at a time and kick off a reindex for all of them, but a user could in theory open each index that needed to be reindexed and kick off the process consecutively.